build-system/babel-plugins/babel-plugin-transform-function-declarations/index.js

```diff
@@ -14,119 +14,198 @@
+   * @param {*} path
+   * @return {boolean}
+   */
+  function containsThisExpression(path) {
+    let containsThis = false;
+    path.traverse({
+      ThisExpression() {
+        containsThis = true;
+        path.stop();
+      },
+    });
+    return containsThis;
+  }
 
-  // Making an existing FunctionDeclaration a let VariableDeclaration is safe.
-  // The name was already reserved for the FunctionDeclaration.
-  function createVariableDeclaration(path) {
-    const {params, body, async, id} = t.cloneNode(path.node);
-    const arrowFunction = t.arrowFunctionExpression(
-      params,
-      body.body[0].argument,
-      async
+  /**
+   * If the FunctionDeclaration identifier is used a manner besides a CallExpression, bail.
+   * @param {*} path
+   * @param {string} name
+   * @return {boolean}
+   */
+  function referencesAreOnlyCallExpressions(path, name) {
+    const binding = path.scope.getBinding(name);
+    return !binding.referencePaths.every((p) =>
+      p.parentPath.isCallExpression({callee: p.node})
     );
-    const declarations = t.variableDeclarator(
-      t.identifier(id.name),
-      arrowFunction
-    );
-    const declaration = t.variableDeclaration('let', [declarations]);
+  }
+
+  function createArrowFunctionExpression(node) {
+    const {params, body, async} = t.cloneNode(node);
+    return t.arrowFunctionExpression(params, body.body[0].argument, async);
+  }
+
+  function createVariableDeclarator(path, name) {
+    const arrowFunction = createArrowFunctionExpression(path.node);
+    return t.variableDeclarator(t.identifier(name), arrowFunction);
+  }
+
+  function createVariableDeclaration(path) {
+    const declarator = createVariableDeclarator(path, path.node.id.name);
+    const declaration = t.variableDeclaration('let', [declarator]);
 
     t.inherits(declaration, path.node);
     return declaration;
   }
 
+  const DECLARATION_BAIL_OUT_CONDITIONS = [
+    isNotFunction,
+    isGenerator,
+    isNotSingleReturnStatement,
+    containsArgumentsUsage,
+    containsThisExpression,
+    (path) => referencesAreOnlyCallExpressions(path, path.get('id').node.name),
+  ];
+
+  // If CallExpression names should be exhempt from arrow conversion, denote them here.
+  const EXEMPTED_EXPRESSION_NAME_REGEXS = [/registerService/];
+
+  const EXPRESSION_BAIL_OUT_CONDITIONS = [
+    isNotFunction,
+    isGenerator,
+    isNotSingleReturnStatement,
+    containsArgumentsUsage,
+    containsThisExpression,
+    (path) => {
+      const callExpression = path.findParent((p) => p.isCallExpression());
+      if (callExpression) {
+        const {name} = (callExpression.node && callExpression.node.callee) || {
+          name: null,
+        };
+        return EXEMPTED_EXPRESSION_NAME_REGEXS.every((regexp) =>
+          regexp.test(name)
+        );
+      }
+
+      const declarator = path.findParent((p) => p.isVariableDeclarator());
```

Comment:
This will traverse a bit too far. Eg, `let foo = call(function() {})`. Here, the declarator isn't for the function expression, it's for the call result.

What I think needs to happen here is to check if the direct parent is a variable declarator. If so, then grab its id.